zerostash-files/src/tree.rs

```diff
@@ -0,0 +1,674 @@
+                            to_remove.push(path);
+                        } else {
+                            entries.scan(|relative_path, noderef| {
+                                let full_path = if path.is_empty() {
+                                    relative_path.clone()
+                                } else {
+                                    format!("{}/{}", path, relative_path)
+                                };
+                                stack.push((full_path, *noderef));
+                            });
+                        }
+                    }
+                }
+            }
+        }
+
+        for key in to_remove {
+            _ = self.remove(&key);
+        }
+    }
+
+    pub fn iter_files(&self) -> TreeIterator {
+        self.insert_root().unwrap();
+        let root = Arc::clone(&self.root());
+        let stack = scc::Stack::default();
+        stack.push((String::new(), root));
+        TreeIterator { stack, inner: self }
+    }
+}
+
+pub struct TreeIterator<'a> {
+    stack: scc::Stack<(String, Arc<Node>)>,
+    inner: &'a Tree,
+}
+
+impl<'a> Iterator for TreeIterator<'a> {
+    type Item = (String, Arc<Entry>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while let Some(next) = self.stack.pop() {
+            let (prefix, node) = (next.0.to_string(), next.1.as_ref());
+            match node {
+                Node::File { refs: _, entry } => return Some((prefix, entry.clone())),
+                Node::Directory { entries } => {
+                    entries.scan(|name, digest| {
+                        let path = if prefix.is_empty() {
+                            name.to_string()
+                        } else {
+                            format!("{prefix}/{name}")
+                        };
+                        if let Some(curr) = self.inner.0.get(digest) {
+                            self.stack.push((path, curr));
+                        }
+                    });
+                }
+            }
+        }
+        None
+    }
+}
+
+impl Collection for Tree {
+    type Depth = infinitree::fields::depth::Incremental;
+
+    type Key = <InnerTree as Collection>::Key;
+
+    type Serialized = <InnerTree as Collection>::Serialized;
+
+    type Item = <InnerTree as Collection>::Item;
+
+    fn key(from: &Self::Serialized) -> &Self::Key {
+        <InnerTree as Collection>::key(from)
+    }
+
+    fn load(from: Self::Serialized, object: &mut dyn infinitree::object::Reader) -> Self::Item {
+        <InnerTree as Collection>::load(from, object)
+    }
+
+    fn insert(&mut self, record: Self::Item) {
+        <InnerTree as Collection>::insert(&mut self.0, record)
+    }
+}
+
+impl Store for Tree {
+    fn store(
+        &mut self,
+        transaction: &mut dyn infinitree::index::Transaction,
+        object: &mut dyn infinitree::object::Writer,
+    ) {
+        <InnerTree as Store>::store(&mut self.0, transaction, object)
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
These look good. Can you add a test for updating a file and verifying it has the correct contents after update, potentially with a full commit/re-open cycle?